src/Core/Services/Implementations/OrganizationService.cs

```diff
@@ -978,6 +978,114 @@ public async Task DisableTwoFactorProviderAsync(Organization organization, TwoFa
             await UpdateAsync(organization);
         }
 
+        private async Task<List<OrganizationUser>> InviteUsersAsync(Guid organizationId, Guid? invitingUserId, IEnumerable<(OrganizationUserInvite, string)> invites)
+        {
+            var organization = await GetOrgById(organizationId);
+            if (organization == null || invites.Any(i => i.Item1.Emails == null || i.Item2 == null))
+            {
+                throw new NotFoundException();
+            }
+
+            var inviteTypes = new HashSet<OrganizationUserType>(invites.Where(i => i.Item1.Type.HasValue).Select(i => i.Item1.Type.Value));
+            if (invitingUserId.HasValue && inviteTypes.Count > 0)
+            {
+                foreach(var type in inviteTypes)
+                {
+                    await ValidateOrganizationUserUpdatePermissions(invitingUserId.Value, organizationId, type, null);
+                }
+            }
+
+            if (organization.Seats.HasValue)
+            {
+                var userCount = await _organizationUserRepository.GetCountByOrganizationIdAsync(organizationId);
+                var availableSeats = organization.Seats.Value - userCount;
+                if (availableSeats < invites.Select(i => i.Item1.Emails.Count()).Sum())
+                {
+                    throw new BadRequestException("You have reached the maximum number of users " +
+                        $"({organization.Seats.Value}) for this organization.");
+                }
+            }
+
+            var orgUsers = new List<OrganizationUser>();
+            var orgUserInvitedCount = 0;
+            var exceptions = new List<Exception>();
+            var events = new List<(OrganizationUser, EventType, DateTime?)>();
+            var existingEmails = new HashSet<string>(await _organizationUserRepository.SelectKnownEmailsAsync(
+                organizationId, invites.SelectMany(i => i.Item1.Emails), false));
+            foreach(var (invite, externalId) in invites)
+            {
+                foreach(var email in invite.Emails)
+                {
+                    try
+                    {
+                        // Make sure user is not already invited
+                        if (existingEmails.Contains(email))
+                        {
+                            continue;
+                        }
+
+                        var orgUser = new OrganizationUser
+                        {
+                            OrganizationId = organizationId,
+                            UserId = null,
+                            Email = email.ToLowerInvariant(),
```

Comment:
So I noticed here you're lower-casing the email address, which is great, but you're using a `HashSet<T>` above for existing emails and other functions earlier in this method w/o doing this... perhaps it would be prudent to create a full `List<string>` of emails or something like that which already transforms the list using `ToLowerInvariant()` and perhaps is even deduplicated (since this is a bulk action).

Reply:
Good idea. I'm changing the HashSet comparitor to `String.InvariantCultureIgnoreCase`